### PR TITLE
Adding support for DPoP.

### LINKED ISSUE #140: Use DPoP authentication scheme for issuer's endpoints
[RFC9449](https://datatracker.ietf.org/doc/html/rfc9449) defines DPoP authorization scheme.
It should be an option for protecting issuer's endpoints 

- [x] Replace Bearer Authentication scheme with DPoP. Spring security doesn't support this
- [x] Enable Keycloak docker image DPoP feature
- [x] Find an alternative way of accessing user claims from Keycloak (used as a source for PID credentials). Currently, issuer just uses the bearer `access_token` send by the wallet to access userinfo endpoint

### PR DESCRIPTION
**Pull request by dzarras**
Closes #140 

The PR protects the credential issuer endpoints with bearer authentication scheme (this was the case so far).
In addition, during boot strapping, application checks whether the OAUTH2 server has been configured to support DPoP or not.
If DPoP is supported, application will also register DPoP authentication scheme.

To sum up, Credential Issuer:
- Always support bearer authentication 
- Conditionally, can accept DPoP access tokens depending the configuration of OAUTH2 server
- The usage of DPoP(or not) is controlled only by OUATH2 server 
- Issuer obtains user claims from OAUTH2 using Keycloak client. This replaced access to userinfo endpoint since it cannot be used with DPoP.

### CODE REVIEW

**Comment by dzarras** [comment]
Currently mDL issuance is working as expected using DPoP.

PID issuance is currently not working because we can no longer use the userinfo endpoint to fetch the user's details.
Currently investigating keycloak-admin client to see if it's a viable alternative.

**Comment by dzarras** [comment]
GetPidFromAuthServer has been implemented using Keycloak admin client. PID issuance has been reinstated.
Leaving this as a draft to better check on Monday whether we are missing anything else.

**Comment by babisRoutis** [comment]
@dzarras Please rebase with main

File: src/main/kotlin/eu/europa/ec/eudi/pidissuer/port/input/AuthorizationContext.kt [hunk 1]
```diff
@@ -18,10 +18,13 @@ package eu.europa.ec.eudi.pidissuer.port.input
 import arrow.core.NonEmptySet
 import eu.europa.ec.eudi.pidissuer.domain.Scope
 
+typealias Username = String
+typealias AccessToken = String
```
Comment (babisRoutis) [inline src/main/kotlin/eu/europa/ec/eudi/pidissuer/port/input/AuthorizationContext.kt:22]: Consider using the base class `AccessToken` from Nimbus

File: src/main/kotlin/eu/europa/ec/eudi/pidissuer/security/DPoPConfigurationProperties.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import arrow.core.NonEmptySet
+import arrow.core.toNonEmptySetOrNull
+import com.nimbusds.jose.JWSAlgorithm
+import org.springframework.boot.context.properties.ConfigurationProperties
+import java.time.Duration
+
+/**
+ * Properties for configuring DPoP.
+ */
+@ConfigurationProperties("spring.security.oauth2.resourceserver.dpop")
```
Comment (babisRoutis) [inline src/main/kotlin/eu/europa/ec/eudi/pidissuer/security/DPoPConfigurationProperties.kt:27]: Why use annotations?

File: src/main/kotlin/eu/europa/ec/eudi/pidissuer/security/DPoPTokenReactiveAuthenticationManager.kt [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+class DPoPTokenReactiveAuthenticationManager(
+    private val introspector: SpringReactiveOpaqueTokenIntrospector,
+    private val verifier: DPoPProtectedResourceRequestVerifier,
+) : ReactiveAuthenticationManager {
+
+    /**
+     * Performs DPoP authentication.
+     */
+    override fun authenticate(authentication: Authentication): Mono<Authentication> =
+        Mono.just(authentication)
```
Comment (babisRoutis) [inline src/main/kotlin/eu/europa/ec/eudi/pidissuer/security/DPoPTokenReactiveAuthenticationManager.kt:51]: Can we remove the Mono implementation of the method?
Instead perhaps we can use the `mono()` method that accepts a suspended lamda

File: src/main/kotlin/eu/europa/ec/eudi/pidissuer/security/DPoPTokenReactiveAuthenticationManager.kt [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    /**
+     * Given a [DPoPTokenAuthentication], the [DPoPIssuer], and the [JWKThumbprintConfirmation], performs DPoP verification
+     * using [verifier].
+     */
+    private fun verify(
+        authentication: DPoPTokenAuthentication,
+        issuer: DPoPIssuer,
+        thumbprint: JWKThumbprintConfirmation,
+    ): Mono<Unit> =
+        Mono.fromCallable {
```
Comment (babisRoutis) [inline src/main/kotlin/eu/europa/ec/eudi/pidissuer/security/DPoPTokenReactiveAuthenticationManager.kt:91]: Can we remove the Mono from implementation?
Use `mono()` which lifts a suspended lamda into Mono context instead.

File: src/main/kotlin/eu/europa/ec/eudi/pidissuer/security/DPoPTokenReactiveAuthenticationManager.kt [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                }
+            OAuth2AuthenticationException(error, exception)
+        }
+}
+
+/**
+ * Gets the [DPoPIssuer] from this [OAuth2AuthenticatedPrincipal].
+ */
+private fun OAuth2AuthenticatedPrincipal.issuer(): Mono<DPoPIssuer> =
+    Mono.justOrEmpty(attributes[OAuth2TokenIntrospectionClaimNames.CLIENT_ID])
```
Comment (babisRoutis) [inline src/main/kotlin/eu/europa/ec/eudi/pidissuer/security/DPoPTokenReactiveAuthenticationManager.kt:116]: Can we use `mono()` with a suspended lamda?

File: src/main/kotlin/eu/europa/ec/eudi/pidissuer/security/DPoPTokenReactiveAuthenticationManager.kt [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        .switchIfEmpty {
+            val error = DPoPTokenError.invalidToken("Unable to determine DPoP issuer")
+            Mono.error(OAuth2AuthenticationException(error))
+        }
+
+/**
+ * Gets the [JWKThumbprintConfirmation] from this [OAuth2AuthenticatedPrincipal].
+ */
+private fun OAuth2AuthenticatedPrincipal.jwkThumbprint(): Mono<JWKThumbprintConfirmation> =
+    Mono.fromCallable { JSONObjectUtils.parse(JSONObject.toJSONString(attributes.filterKeys { it == "cnf" })) }
```
Comment (babisRoutis) [inline src/main/kotlin/eu/europa/ec/eudi/pidissuer/security/DPoPTokenReactiveAuthenticationManager.kt:128]: Can we use `mono()` with a suspended lamda?

**Review by babisRoutis — CHANGES_REQUESTED** [review]

**Comment by babisRoutis** [comment]
@dzarras Removed comments related to implementation of methods using `Mono` data structure. We can deal with it after merge.

Please let me know as soon as you have prepared the conditional registration of DPoP authentication scheme

File: src/main/kotlin/eu/europa/ec/eudi/pidissuer/security/DPoPConfigurationProperties.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+ */
+data class DPoPConfigurationProperties(
+    val algorithms: Set<JWSAlgorithm>,
+    val proofMaxAge: Duration,
+    val cachePurgeInterval: Duration,
+    val realm: String?,
+) {
+    init {
+        require(JWSAlgorithm.Family.SIGNATURE.containsAll(algorithms)) { "'algorithms' contains invalid values" }
+        require(!proofMaxAge.isZero) { "'proofMaxAge' cannot be zero" }
```
Comment (babisRoutis) [inline src/main/kotlin/eu/europa/ec/eudi/pidissuer/security/DPoPConfigurationProperties.kt:32]: I have the impression that there could be negative durations. 
Can you please double check this

**Review by babisRoutis — DISMISSED** [review]

**Review by babisRoutis — APPROVED** [review]

**Pull request merged**